src/foolscap/storage.py

```diff
@@ -27,7 +28,36 @@
 from foolscap.slicers.dict import OrderedDictSlicer
 from foolscap.slicers.root import ScopedRootSlicer, ScopedRootUnslicer
 
-
+# PY3KPORT
+try:
+  from types import InstanceType
+except ImportError:
+  InstanceType = object
+
+try:
+  from types import ClassType
+except ImportError:
+  ClassType = type
+
+# PY3KPORT
+def get_class_that_defined_method(meth):
+  """ Function that return the class that defined a given method """
+
+  # Fix for Python3 to get classes for unbound methods
+  # for the MethodSlicer class
+  # Ref: https://stackoverflow.com/questions/3589311/get-defining-class-of-unbound-method-object-in-python-3
+  # with some fixes and simplifications
+  
+  if inspect.ismethod(meth):
+    # If its a method, we can rely on __self__
+    return meth.__self__.__class__
+
+  if inspect.isfunction(meth):
+    cls = getattr(inspect.getmodule(meth),
+                  meth.__qualname__.split('.<locals>', 1)[0].rsplit('.', 1)[0])
+    if isinstance(cls, type):
+      return cls
+  
```

Comment:
whitespace. and using string-splitting to get .. classes? makes me a little nervous. This needs tests, at minimum.